### PR TITLE
Add maxLength in proto for AuthConfigs.configs

### PR DESCRIPTION
**Pull request by arianaw66**
# Description

Please include a high level summary of the changes.

This bug fixes ... \ This new feature can be used to ...

_Fill out any of the following sections that are relevant and remove the others_
## API changes
- Added x field to y resource
- ...

## Code changes
- Fix error in `Foo()` function
- Add `Bar()` function
- ...

## CI changes
- Adjusted schedule for x job
- ...

## Docs changes
- Added guide about feature x to public docs
- Updated README to account for y behavior
- ...

# Context

Users ran into this bug doing ... \ Users needed this feature to ...

See slack conversation [here](https://solo-io-corp.slack.com/archives/some/post)

## Interesting decisions
 
We chose to do things this way because ...

## Testing steps

I manually verified behavior by ...

## Notes for reviewers

Be sure to verify intended behavior by ...

Please proofread comments on ...

This is a complex PR and may require a huddle to discuss ...

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto.sk.md [hunk 1]
```diff
@@ -164,7 +164,7 @@ format that will be included in the extauth snapshot.
...
-| `configs` | [[]enterprise.gloo.solo.io.AuthConfig.Config](../extauth.proto.sk/#config) | List of auth configs to be checked for requests on a route referencing this auth config, By default, every config must be authorized for the entire request to be authorized. This behavior can be changed by defining names for each config and defining `boolean_expr` below. State is shared between successful requests on the chain, i.e., the headers returned from each successful auth service get appended into the final auth response. |
+| `configs` | [[]enterprise.gloo.solo.io.AuthConfig.Config](../extauth.proto.sk/#config) | List of auth configs to be checked for requests on a route referencing this auth config, By default, every config must be authorized for the entire request to be authorized. This behavior can be changed by defining names for each config and defining `boolean_expr` below. State is shared between successful requests on the chain, i.e., the headers returned from each successful auth service get appended into the final auth response. +kubebuilder:validation:MaxLength=100000. |
```
Comment (arianaw66) [inline docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto.sk.md:167]: Noting that docsgen doesn't hide this annotation

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 2]
```diff
@@ -353,6 +364,9 @@ message OAuth {
 
     // we to redirect after successful auth, if we can't determine the original
     // url this should be your publicly available app url.
+    //
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
     string app_url = 4 [deprecated = true];
```
Comment (npolshakova) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:370]: Do we need validation here if `[deprecated = true]` is set?

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 3]
```diff
@@ -45,6 +45,8 @@ message AuthConfig {
         google.protobuf.StringValue name = 9;
 
         oneof auth_config {
+            // +kubebuilder:validation:XValidation:rule="has(self.apr) || (has(self.encryption) && has(self.userList))",message="Either apr or both encryption and userSource must be set"
+            // +kubebuilder:validation:XValidation:rule="has(self.apr) ? !has(self.encryption) && !has(self.userList) : true",message="If apr is set, neither encryption nor userSource may be set"
```
Comment (npolshakova) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:49]: Do we want to combine these into one rule? Or are they separate for readability? I think you could do something like?

```
has(self.apr) ? (!has(self.encryption) && !has(self.userList)) : (has(self.encryption) && has(self.userList))
```

The message would be a little longer though- either apr must be set, or both encryption and userSource must be set, but not both apr and encryption/userSource simultaneously.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 4]
```diff
@@ -368,6 +382,8 @@ message OAuth2 {
         // provide issuer location and let gloo handle OIDC flow for you.
         // requests authorized by validating the contents of ID token.
         // can also authorize the access token if configured.
+        //
+        // +kubebuilder:validation:XValidation:rule="has(self.clientAuthentication) ? !has(self.clientSecretRef) && !has(self.disableClientSecret) : has(self.clientSecretRef) || (has(self.disableClientSecret) && self.disableClientSecret)",message="If clientAuthentication is set, neither clientSecretRef nor disableClientSecret may be set. Otherwise, clientSecretRef must be set or disableClientSecret must be true."
```
Comment (npolshakova) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:386]: Are these enforced as non empty somewhere else? https://github.com/solo-io/solo-projects/blob/a1b7f14e99578ea98ab7b9bcdbe42a6c8f70a8d8/projects/gloo/pkg/syncer/extauth/validate.go#L114

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 5]
```diff
@@ -969,6 +1028,8 @@ message AccessTokenValidation {
         // If provided, the (opaque) access token provided or received from the oauth authorization endpoint
         // will be validated against this endpoint, or locally cached responses for this access token.
         // This field is deprecated as it does not support authenticated introspection requests
+        //
+        // +kubebuilder:validation:MinLength=1
```
Comment (npolshakova) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1032]: Do we need this if it's deprecated?

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 6]
```diff
@@ -939,6 +995,9 @@ message IntrospectionValidation {
     // The URL for the [OAuth2.0 Token Introspection](https://tools.ietf.org/html/rfc7662) endpoint.
     // If provided, the (opaque) access token provided or received from the oauth authorization endpoint
     // will be validated against this endpoint, or locally cached responses for this access token.
+    //
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
     string introspection_url = 1;
```
Comment (npolshakova) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1001]: Is `IsIntrospectionUrlParsable` going to be a CEL++ rule? https://github.com/solo-io/solo-projects/blob/a1b7f14e99578ea98ab7b9bcdbe42a6c8f70a8d8/projects/gloo/pkg/syncer/extauth/validate.go#L165C17-L165C43

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 7]
```diff
@@ -939,6 +995,9 @@ message IntrospectionValidation {
...
 
     // Your client id as registered with the issuer.
```
Comment (npolshakova) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1003]: Do we need validation for `client_id` and `client_secret_ref`? We check they are not empty in the [code](https://github.com/solo-io/solo-projects/blob/a1b7f14e99578ea98ab7b9bcdbe42a6c8f70a8d8/projects/gloo/pkg/syncer/extauth/validate.go#L171), but are the CEL rules enforced somewhere else?

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 8]
```diff
@@ -45,6 +45,8 @@ message AuthConfig {
         google.protobuf.StringValue name = 9;
 
         oneof auth_config {
```
[the remark this answers is not in the document]
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:47]: Good question. Edge doesn't have snapshots, right? Does it have some equivalent?
Also I'm not sure how the conformance tests work and how easy it would be to fit Edge resources into that framework but maybe @shashankram has thoughts on this.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 2 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:370]: The entire `oauth` authConfig type is deprecated, but this is not enforced by `kubectl apply` in any way I'm aware of. If the `oauth` type is used anyway, imo the validations should still be enforced on subfields.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 5 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1032]: I don't really see why not. Fwiw, this corresponds to [this line](https://github.com/solo-io/solo-projects/blob/a1b7f14e99578ea98ab7b9bcdbe42a6c8f70a8d8/projects/gloo/pkg/syncer/extauth/validate.go#L156) in the code and would catch pre-admission instead of post.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 4 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:386]: You mean appUrl, clientId, etc.? Yes, required fields are enforced on the fields themselves (e.g. [here](https://github.com/solo-io/gloo/pull/9481/files/97dca251ccd93e327eb95973c72630b1ab9e5480#diff-9a3a741756fd9cf3b5fbaceb118314b2ef89bb148958533e4cad599153d70e29R637-R638)) while rules that take into account multiple fields need to be defined at the level of the parent.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 6 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1001]: Yeah, I think that's the right call

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 7 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1003]: The code you're referencing doesn't error if those are non-empty. It's a more complex rule that's enforced in the line you commented on [here](https://github.com/solo-io/gloo/pull/9481#discussion_r1618938620).

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 3 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:49]: Yeah, we can certainly combine these. I'm conflicted about readability but the docs folks can also help with the wording.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 9]
```diff
@@ -776,6 +805,8 @@ message OidcAuthorizationCode {
          // Private Key JWT Authentication requires a signing key for the JWT and an duration for the JWT to be valid.
         message PrivateKeyJwt{
             // Signing key for the JWT used to authenticate the client
+            //
```
Comment (dmitri-d) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:808]: Should ClientSecret have validation added to it too (here: https://github.com/solo-io/gloo/blob/59b5013a26e8c3aec4285f805717a847e8d2fee2/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto#L798). Not sure if that was in scope.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 10]
```diff
@@ -785,12 +816,15 @@ message OidcAuthorizationCode {
...
             ClientSecret client_secret = 1;
             // Use the private ket JWT method to authenticate the client
             PrivateKeyJwt private_key_jwt = 2;
         }
     }
 
+    // +kubebuilder:validation:XValidation:rule="has(self.clientSecret) || has(self.privateKeyJwt)",message="Must specify clientSecret or privateKeyJwt"
     ClientAuthentication client_authentication = 25;
 
     oneof Provider {
```
Comment (dmitri-d) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:830]: Also not sure if Azure provider validation was in scope, but could be validated too: if any of client_id, tenant_id, and client_secret are set, then all of them must be set...

**Review by npolshakova — APPROVED** [review]

**Comment by shashankram** [comment]
Since this API change effects Gloo Edge users, has it been approved by the relevant stakeholders? I know Field has some concerns with enabling such validations by default, for which we've had discussions around building a toggle in CRDs to conditionally not render these validations. 

I'd like to see a signoff from @sam-heilbron @DuncanDoyle and @kcbabo  before this merges.

**Comment by sam-heilbron** [comment]
I just added the `work in progress` label to prevent this from merging right away. I like the proposed changes a lot. Given that this is late in our release cycle for 1.17, and introduces changes to our validation pipeline, I want to make sure we have adequate testing for the feature and more people on the Gloo Gateway team understand the changes. I'm going to sync with product to understand the urgency of this work being introduced. My current vote is that we hold off on introducing this until after 1.17 branch is cut. That way it can soak in the main branch for 1.18 for a while. cc @arianaw66

**Comment by htpvu** [comment]
> Since this API change effects Gloo Edge users, has it been approved by the relevant stakeholders? I know Field has some concerns with enabling such validations by default, for which we've had discussions around building a toggle in CRDs to conditionally not render these validations.
> 
> I'd like to see a signoff from @sam-heilbron @DuncanDoyle and @kcbabo before this merges.

I've brought this change to the attention of @DuncanDoyle @kcbabo and @sam-heilbron, we have their thumbs up on the change and we can merge this PR right after the 1.17 branch is cut.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 9 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:808]: If I understand correctly which constraint you're suggesting, it's added [here](https://github.com/solo-io/gloo/blob/AuthConfig.configs-maxLength/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto#L820) on the parent, since it requires access to both the client_secret_ref field and disable_client_secret.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 10 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:830]: Everything's in scope, but can you tell me where this constraint comes from? Just something you know of from your OIDC knowledge?

Also, do you know what the behavior is in both Edge and GME if the constraint is not followed? I want to confirm the behavior in Edge in particular, since Edge does some massaging of inputs before sending to the ext-auth-service, so we need to be careful to not break any valid Edge setups with these changes, even if they are considered invalid by GME.

Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:830]: Based on the ext-auth-service code, seems like all 3 are individually required if the Azure field is set:
https://github.com/solo-io/ext-auth-service/blob/v0.56.8/pkg/controller/translation/oauth.go#L187-L195

Comment (dmitri-d) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:830]: > but can you tell me where this constraint comes from?

There's a couple of spots where we make a check if Azure OIDC provider is configured (here's the check itself https://github.com/solo-io/ext-auth-service/blob/master/pkg/config/oidc/config.go#L130). I think it would be a better interface if we validated if all/none of client_id, tenant_id, and client_secret are set.

> do you know what the behavior is in both Edge and GME if the constraint is not followed?

I'm not sure if edge exposes or supports OIDC config, tbh. The work Tyler and I did was in GM.

Comment (dmitri-d) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:830]: It all makes sense; there are runtime checks in place, and while I think that having validations would result in clearer UI/fewer surprises, it's by no means a deal breaker. tyvm for looking into it!

**Comment by arianaw66** [comment]
It appears the bump to solo-kit v0.35.1 breaks the build-bot tests
See https://github.com/solo-io/gloo/pull/9626

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 8 above)
[the remark this answers is not in the document]
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:47]: This is still a valid question though less pressing now that only required fields will be enforced in Edge.

File: docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto.sk.md [hunk 11]
```diff
@@ -164,7 +164,7 @@ format that will be included in the extauth snapshot.
...
-| `configs` | [[]enterprise.gloo.solo.io.AuthConfig.Config](../extauth.proto.sk/#config) | List of auth configs to be checked for requests on a route referencing this auth config, By default, every config must be authorized for the entire request to be authorized. This behavior can be changed by defining names for each config and defining `boolean_expr` below. State is shared between successful requests on the chain, i.e., the headers returned from each successful auth service get appended into the final auth response. |
+| `configs` | [[]enterprise.gloo.solo.io.AuthConfig.Config](../extauth.proto.sk/#config) | List of auth configs to be checked for requests on a route referencing this auth config, By default, every config must be authorized for the entire request to be authorized. This behavior can be changed by defining names for each config and defining `boolean_expr` below. State is shared between successful requests on the chain, i.e., the headers returned from each successful auth service get appended into the final auth response. +kubebuilder:validation:Required +kubebuilder:validation:MinItems=1. |
```
Comment (Rachael-Graham) [inline docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto.sk.md:167]: I'm not sure why (I'm far less familiar with the api doc generation in the edge repo) but the kubebuilder comments are being built directly into the docs, whereas in the gme repo they were being ignored. hmm @arianaw66 do you remember if you (or @nikolasmatt ) added a rule in gme to ignore these somehow?

Comment (arianaw66) [reply docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto.sk.md:167]: Yeah I called that out [above](https://github.com/solo-io/gloo/pull/9481#discussion_r1600224659). [According to Nadine](https://solo-io-corp.slack.com/archives/CEDCS8TAP/p1716234451569969?thread_ts=1716234094.464939&cid=CEDCS8TAP), we don't have that functionality in gloo. I'm not sure how hard it would be to add it to solo-kit similarly to how it was added to skv2- maybe @shashankram would know?

**Review by dmitri-d — APPROVED** [review]

**Review by Rachael-Graham — APPROVED** [review]
Approving for now. There's some followup doc work I'd like to do, which will include 1) translating rules into user-friendly comments, and 2) figuring out some way to update solo-kit so that the cel rules themselves dont build into the published docs.

**Review by npolshakova — APPROVED** [review]

**Review by sheidkamp — COMMENTED** [review]
"Approved"

- Checked the rules against the `CheckIfInvalidAuthConfig` function
- Pulled changes into `solo-projects` and manually ran the in memory and kubernetes e2e ext-auth tests 
- Ran ./projects/gloo/pkg/syncer/extauth validation unit tests

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 6 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1000]: Yes, it's our shorthand for any Go validation in our translator that checks constraints that are too complex to be checked by CEL on admission

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 10 above)
Comment (arianaw66) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:830]: To follow up on this, I can't speak to actual auth behavior, but in terms of the AuthConfig CR's status:
- GME seems to require all of these fields, even if none is set
- Edge seems to require none of these fields and doesn't really seem to do any validation on the `azure` section
